### PR TITLE
Unlock snapshots in import-db mode + AccountsDB constructor refactor

### PR DESCRIPTION
**Pull request by iulianpascalau**
- call `SetIdle` when trying to do a snapshot when running in import-db
- refactored AccountsDB constructor to use arguments

### CODE REVIEW

File: state/export_test.go [hunk 1]
```diff
@@ -2,36 +2,49 @@ package state
...
 func (adb *AccountsDB) GetAccount(address []byte) (vmcommon.AccountHandler, error) {
 	return adb.getAccount(address)
 }
 
+// GetObsoleteHashes -
 func (adb *AccountsDB) GetObsoleteHashes() map[string][][]byte {
 	return adb.obsoleteDataTrieHashes
 }
 
+// GetObsoleteHashes -
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/export_test.go]: done

File: state/peerAccountsDB_test.go [hunk 2]
```diff
@@ -12,127 +12,110 @@ import (
...
+		args.Hasher = nil
 
-	assert.True(t, check.IfNil(adb))
-	assert.Equal(t, state.ErrNilHasher, err)
-}
+		adb, err := state.NewPeerAccountsDB(args)
+		assert.True(t, check.IfNil(adb))
+		assert.Equal(t, state.ErrNilHasher, err)
+	})
+	t.Run("nil hasher should error", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply state/peerAccountsDB_test.go]: done

File: state/accountsDB_test.go [hunk 3]
```diff
@@ -36,19 +36,31 @@ import (
...
+		ProcessStatusHandler:  &testscommon.ProcessStatusHandlerStub{},
+	}
+}
+
+func generateAccountDBFromTrie(trie common.Trie) *state.AccountsDB {
+	args := createMockAccountsDBArgs()
+	args.Trie = trie
+
+	accnt, _ := state.NewAccountsDB(args)
 	return accnt
```
Comment (dragos-rebegea) [inline state/accountsDB_test.go:64]:
```suggestion
	accounts, _ := state.NewAccountsDB(args)
	return accounts
```

File: state/accountsDB_test.go [hunk 4]
```diff
@@ -1621,9 +1590,20 @@ func TestAccountsDB_MainTrieAutomaticallyMarksCodeUpdatesForEviction(t *testing.
 	}
 	storageManager, _ := trie.NewTrieStorageManager(args)
 	maxTrieLevelInMemory := uint(5)
-	tr, _ := trie.NewTrie(storageManager, marshalizer, hsh, maxTrieLevelInMemory)
+	tr, _ := trie.NewTrie(storageManager, marshaller, hasher, maxTrieLevelInMemory)
 	spm, _ := storagePruningManager.NewStoragePruningManager(ewl, 5)
-	adb, _ := state.NewAccountsDB(tr, hsh, marshalizer, factory.NewAccountCreator(), spm, common.Normal)
+
+	argsAccountsDB := state.ArgsAccountsDB{
```
Comment (dragos-rebegea) [inline state/accountsDB_test.go:1596]: use createMockAccountsDBArgs

File: state/accountsDB_test.go [hunk 5]
```diff
@@ -1683,26 +1663,36 @@ func TestAccountsDB_RemoveAccountMarksObsoleteHashesForEviction(t *testing.T) {
...
 		EpochNotifier:          &epochNotifier.EpochNotifierStub{},
 		IdleProvider:           &testscommon.ProcessStatusHandlerStub{},
 	}
 	storageManager, _ := trie.NewTrieStorageManager(args)
-	tr, _ := trie.NewTrie(storageManager, marshalizer, hsh, maxTrieLevelInMemory)
+	tr, _ := trie.NewTrie(storageManager, marshaller, hasher, maxTrieLevelInMemory)
 	spm, _ := storagePruningManager.NewStoragePruningManager(ewl, 5)
-	adb, _ := state.NewAccountsDB(tr, hsh, marshalizer, factory.NewAccountCreator(), spm, common.Normal)
+
+	argsAccountsDB := state.ArgsAccountsDB{
```
Comment (dragos-rebegea) [inline state/accountsDB_test.go:1686]: use createMockAccountsDBArgs

File: state/storagePruningManager/storagePruningManager_test.go [hunk 6]
```diff
@@ -40,10 +40,20 @@ func getDefaultTrieAndAccountsDbAndStoragePruningManager() (common.Trie, *state.
...
 	}
 	trieStorage, _ := trie.NewTrieStorageManager(args)
-	tr, _ := trie.NewTrie(trieStorage, marshalizer, hsh, 5)
-	ewl, _ := evictionWaitingList.NewEvictionWaitingList(100, testscommon.NewMemDbMock(), marshalizer)
+	tr, _ := trie.NewTrie(trieStorage, marshaller, hasher, 5)
+	ewl, _ := evictionWaitingList.NewEvictionWaitingList(100, testscommon.NewMemDbMock(), marshaller)
 	spm, _ := NewStoragePruningManager(ewl, generalCfg.PruningBufferLen)
-	adb, _ := state.NewAccountsDB(tr, hsh, marshalizer, factory.NewAccountCreator(), spm, common.Normal)
+
+	argsAccountsDB := state.ArgsAccountsDB{
```
Comment (dragos-rebegea) [inline state/storagePruningManager/storagePruningManager_test.go:47]: use createMockAccountsDBArgs

File: state/accountsDB_test.go [hunk 7]
```diff
@@ -2346,9 +2354,18 @@ func BenchmarkAccountsDb_GetCodeEntry(b *testing.B) {
 		IdleProvider:           &testscommon.ProcessStatusHandlerStub{},
 	}
 	storageManager, _ := trie.NewTrieStorageManager(args)
-	tr, _ := trie.NewTrie(storageManager, marshalizer, hasher, maxTrieLevelInMemory)
+	tr, _ := trie.NewTrie(storageManager, marshaller, hasher, maxTrieLevelInMemory)
 	spm := disabled.NewDisabledStoragePruningManager()
-	adb, _ := state.NewAccountsDB(tr, hasher, marshalizer, factory.NewAccountCreator(), spm, common.Normal)
+	argsAccountsDB := state.ArgsAccountsDB{
```
Comment (bogdan-rosianu) [inline state/accountsDB_test.go:2359]: use createMockAccountsDBArgs

File: state/accountsDB_test.go [hunk 8]
```diff
@@ -2229,11 +2228,20 @@ func TestAccountsDB_Close(t *testing.T) {
...
 	}
-	marshalizer := &testscommon.MarshalizerMock{}
-	hsh := &hashingMocks.HasherMock{}
-	ewl, _ := evictionWaitingList.NewEvictionWaitingList(100, testscommon.NewMemDbMock(), marshalizer)
+	marshaller := &testscommon.MarshalizerMock{}
+	hasher := &hashingMocks.HasherMock{}
+	ewl, _ := evictionWaitingList.NewEvictionWaitingList(100, testscommon.NewMemDbMock(), marshaller)
 	spm, _ := storagePruningManager.NewStoragePruningManager(ewl, 10)
-	adb, _ := state.NewAccountsDB(tr, hsh, marshalizer, factory.NewAccountCreator(), spm, common.Normal)
+	argsAccountsDB := state.ArgsAccountsDB{
```
Comment (bogdan-rosianu) [inline state/accountsDB_test.go:2235]: use createMockAccountsDBArgs

File: state/accountsDB_test.go [hunk 9]
```diff
@@ -2123,9 +2113,18 @@ func TestAccountsDB_GetCode(t *testing.T) {
 		IdleProvider:           &testscommon.ProcessStatusHandlerStub{},
 	}
 	storageManager, _ := trie.NewTrieStorageManager(args)
-	tr, _ := trie.NewTrie(storageManager, marshalizer, hasher, maxTrieLevelInMemory)
+	tr, _ := trie.NewTrie(storageManager, marshaller, hasher, maxTrieLevelInMemory)
 	spm := disabled.NewDisabledStoragePruningManager()
-	adb, _ := state.NewAccountsDB(tr, hasher, marshalizer, factory.NewAccountCreator(), spm, common.Normal)
+	argsAccountsDB := state.ArgsAccountsDB{
```
Comment (bogdan-rosianu) [inline state/accountsDB_test.go:2118]: use createMockAccountsDBArgs

File: state/accountsDB_test.go (hunk 3 above)
Comment (iulianpascalau) [reply state/accountsDB_test.go:64]: done

File: state/accountsDB_test.go (hunk 4 above)
Comment (iulianpascalau) [reply state/accountsDB_test.go:1596]: done

File: state/accountsDB_test.go (hunk 5 above)
Comment (iulianpascalau) [reply state/accountsDB_test.go:1686]: done

File: state/storagePruningManager/storagePruningManager_test.go (hunk 6 above)
Comment (iulianpascalau) [reply state/storagePruningManager/storagePruningManager_test.go:47]: can not do this, the createMockAccountsDBArgs is available only in `state` package. No reason to redefine it here

File: genesis/process/memoryComponents.go [hunk 10]
```diff
@@ -12,24 +13,27 @@ import (
...
-		common.Normal,
-	)
+	args := state.ArgsAccountsDB{
+		Trie:                  tr,
+		Hasher:                hasher,
+		Marshaller:            marshaller,
+		AccountFactory:        accountFactory,
+		StoragePruningManager: disabled.NewDisabledStoragePruningManager(),
+		ProcessingMode:        common.Normal,
+		ProcessStatusHandler:  &commonDisabled.ProcessStatusHandler{},
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply genesis/process/memoryComponents.go:33]: done, added unit tests

File: state/accountsDB_test.go (hunk 7 above)
Comment (iulianpascalau) [reply state/accountsDB_test.go:2359]: done

File: state/accountsDB_test.go (hunk 8 above)
Comment (iulianpascalau) [reply state/accountsDB_test.go:2235]: done

File: state/accountsDB_test.go (hunk 9 above)
Comment (iulianpascalau) [reply state/accountsDB_test.go:2118]: done

**Review by bogdan-rosianu — APPROVED** [review]

**Review by dragos-rebegea — APPROVED** [review]

**Review by valentin-lup — APPROVED** [review]
Validated with staggered ImportDB on all shards !

**Review by popenta — APPROVED** [review]
@@                            Log scanner                                  @@
##                            unlock-snapshots-in-import-db                ##
================================================================================
- Known Warnings              425
- New Warnings                283
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================

**Pull request merged**